decomposer.go

```diff
@@ -0,0 +1,114 @@
+// A zero length coefficient is a zero value.
+// If the form is not finite the coefficient and scale should be ignored.
+// The negative parameter may be set to true for any form, although implementations are not required
+// to respect the negative parameter in the non-finite form.
+//
+// Implementations may choose to signal a negative zero or negative NaN, but implementations
+// that do not support these may also ignore the negative zero or negative NaN without error.
+// If an implementation does not support Infinity it may be converted into a NaN without error.
+// If a value is set that is larger then what is supported by an implementation is attempted to
+// be set, an error must be returned.
+// Implementations must return an error if a NaN or Infinity is attempted to be set while neither
+// are supported.
+type decomposer interface {
+	// Decompose returns the internal decimal state into parts.
+	// If the provided buf has sufficient capacity, buf may be returned as the coefficient with
+	// the value set and length set as appropriate.
+	Decompose(buf []byte) (form byte, negative bool, coefficient []byte, exponent int32)
+
+	// Compose sets the internal decimal value from parts. If the value cannot be
+	// represented then an error should be returned.
+	Compose(form byte, negative bool, coefficient []byte, exponent int32) error
+}
+
+// Decompose returns the internal decimal state into parts.
+// If the provided buf has sufficient capacity, buf may be returned as the coefficient with
+// the value set and length set as appropriate.
+func (d *Decimal) Decompose(buf []byte) (form byte, negative bool, coefficient []byte, exponent int32) {
+	switch d.Form {
+	default:
+		panic(fmt.Errorf("unknown Form: %v", d.Form))
+	case Finite:
+		// Nothing, continue on.
+	case Infinite:
+		negative = d.Negative
+		form = 1
+		return
+	case NaNSignaling, NaN:
+		negative = d.Negative
+		form = 2
+		return
+	}
+	// Finite form.
+	negative = d.Negative
+	exponent = d.Exponent
+	coefficient = d.Coeff.Bytes() // Return big-endian, but need little-endian.
+
+	i := 0
+	j := len(coefficient) - 1
+	for i < j {
+		coefficient[i], coefficient[j] = coefficient[j], coefficient[i]
+		i++
+		j--
+	}
+	return
+}
+
+// Compose sets the internal decimal value from parts. If the value cannot be
+// represented then an error should be returned.
+func (d *Decimal) Compose(form byte, negative bool, coefficient []byte, exponent int32) error {
+	switch form {
+	default:
+		return fmt.Errorf("unknown form: %v", form)
+	case 0:
+		d.Form = Finite
+		// Set rest of finite form below.
+	case 1:
+		d.Form = Infinite
+		d.Negative = negative
+		return nil
+	case 2:
+		d.Form = NaN
+		d.Negative = negative
+		return nil
+	}
+	// Finite form.
+	d.Negative = negative
+
+	// coefficient is a little-endian, swap for big-endian.
+	i := 0
+	j := len(coefficient) - 1
+	for i < j {
+		coefficient[i], coefficient[j] = coefficient[j], coefficient[i]
```

Comment:
Is mutating coefficient ok? Feels like this API should document that if so.

Reply:
Great observation. I've updated the [documentation](https://godoc.org/github.com/golang-sql/decomposer#Decimal) to say it is *not* okay to modify the coefficient and updated the implementation here to make a copy first and test for this case.